### PR TITLE
Clean up various DMCompiler build warnings

### PR DESCRIPTION
**Pull request by ike709**
- Mostly just converting CompileErrorExceptions to WarningCode emissions
- Micro-optimized some list/dict allocs
- Everything else should be explained by the commit names

### CODE REVIEW

**Comment by ike709** [comment]
That last commit is kind of ugly and I'll take better suggestions.

File: DMCompiler/DM/Expressions/Constant.cs [hunk 1]
```diff
@@ -414,10 +432,11 @@ private enum PathType {
                 PathType.TypeReference => JsonVariableType.Type,
                 PathType.ProcReference => JsonVariableType.Proc,
                 PathType.ProcStub => JsonVariableType.ProcStub,
-                PathType.VerbStub => JsonVariableType.VerbStub
+                PathType.VerbStub => JsonVariableType.VerbStub,
+                _ => throw new ArgumentOutOfRangeException()
             };
 
-            json = new Dictionary<string, object>() {
+            json = new Dictionary<string, object> {
```
Comment (ike709) [inline DMCompiler/DM/Expressions/Constant.cs:439]:
```suggestion
            json = new Dictionary<string, object>(2) {
```

File: DMCompiler/DM/DMExpression.cs [hunk 2]
```diff
@@ -39,7 +39,8 @@ public enum IdentifierPushResult {
 
         public static void Emit(DMObject dmObject, DMProc proc, DMASTExpression expression, DreamPath? inferredPath = null) {
             var expr = Create(dmObject, proc, expression, inferredPath);
-            expr.EmitPushValue(dmObject, proc);
+            // expr can be null if we error elsewhere
+            expr?.EmitPushValue(dmObject, proc);
```
Comment (willox) [inline DMCompiler/DM/DMExpression.cs:43]: Wouldn't no `EmitPushValue` call happening here due to a null `expr` lead to ill-formed bytecode?

**Comment by willox** [comment]
Same goes for every `?.` call to `EmitPushValue`. Successfully outputting invalid procs shouldn't be a thing

**Comment by ike709** [comment]
image

**Comment by ike709** [comment]
I was hoping to knock out some easy, low hanging fruit today. But this looks like it'll be more involved than that and I'm not in the mood. Might revisit in the future, or (preferably) piecemeal the exception removals.

**Pull request closed without merging**